src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreBuildHandler.cs

```diff
@@ -0,0 +1,172 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Threading;
+using NuGet.Configuration;
+using NuGet.PackageManagement;
+using NuGet.VisualStudio;
+using SystemTask = System.Threading.Tasks.Task;
+using ThreadHelper = Microsoft.VisualStudio.Shell.ThreadHelper;
+
+namespace NuGet.SolutionRestoreManager
+{
+    /// <summary>
+    /// A solution build manager events listener orchestrating on-build restore operations
+    /// </summary>
+    /// <remarks>
+    /// Utilizes four core events to start, monitor, and control restore operations:
+    /// UpdateSolution_Begin
+    /// UpdateSolution_QueryDelayFirstUpdateAction
+    /// UpdateSolution_Cancel
+    /// UpdateSolution_Done
+    /// </remarks>
+    public sealed class SolutionRestoreBuildHandler 
+        : IVsUpdateSolutionEvents5, IDisposable
+    {
+        private const uint VSCOOKIE_NIL = 0;
+
+        [Import]
+        private Lazy<INuGetLockService> LockService { get; set; }
+
+        [Import]
+        private Lazy<ISettings> Settings { get; set; }
+
+        [Import]
+        private Lazy<ISolutionRestoreWorker> SolutionRestoreWorker { get; set; }
+
+        /// <summary>
+        /// The <see cref="IVsSolutionBuildManager3"/> object controlling the update solution events.
+        /// </summary>
+        private IVsSolutionBuildManager3 _solutionBuildManager;
+
+        /// <summary>
+        /// The cookie associated to the the <see cref="IVsUpdateSolutionEvents5"/> events.
+        /// </summary>
+        private uint _updateSolutionEventsCookieEx;
+
+        private SolutionRestoreBuildHandler()
```

Comment:
On second note, This private constructor is needed to construct it's instance inside `InitializeAsync`